ros/src/twist_controller/dbw_node.py

```diff
@@ -53,27 +53,60 @@ def __init__(self):
         self.brake_pub = rospy.Publisher('/vehicle/brake_cmd',
                                          BrakeCmd, queue_size=1)
 
-        # TODO: Create `Controller` object
-        # self.controller = Controller(<Arguments you wish to provide>)
-
-        # TODO: Subscribe to all the topics you need to
+        self.controller = Controller(vehicle_mass=vehicle_mass,
+                                     fuel_capacity=fuel_capacity,
+                                     brake_deadband=brake_deadband,
+                                     decel_limit=decel_limit,
+                                     accel_limit=accel_limit,
+                                     wheel_radius=wheel_radius,
+                                     wheel_base=wheel_base,
+                                     steer_ratio=steer_ratio,
+                                     max_lat_accel=max_lat_accel,
+                                     max_steer_angle=max_steer_angle)
+
+        rospy.loginfo('Staring DBW Node')
+        rospy.Subscriber('/twist_cmd', TwistStamped, self.twist_cb)
+        rospy.Subscriber('/current_velocity', TwistStamped, self.velocity_cb)
+        rospy.Subscriber('/vehicle/dbw_enabled', Bool, self.dbw_enabled_cb)
+
+        self.current_vel = None
+        self.current_ang_vel = None
+        self.dbw_enabled = None
+        self.linear_vel = None
+        self.angular_vel = None
+        self.throttle  = self.steering = self.brake = 0
 
         self.loop()
 
     def loop(self):
         rate = rospy.Rate(50) # 50Hz
         while not rospy.is_shutdown():
-            # TODO: Get predicted throttle, brake, and steering using `twist_controller`
-            # You should only publish the control commands if dbw is enabled
-            # throttle, brake, steering = self.controller.control(<proposed linear velocity>,
-            #                                                     <proposed angular velocity>,
-            #                                                     <current linear velocity>,
-            #                                                     <dbw status>,
-            #                                                     <any other argument you need>)
-            # if <dbw is enabled>:
-            #   self.publish(throttle, brake, steer)
+            if not None in (self.current_vel, self.linear_vel, self.angular_vel):
```

Comment:
I think what you mean is `if None not in (...):`

Reply:
Sorry, I didn't know if there is a difference, let me know if there is one?

`>>> b = [1,2,3]
>>> a = [4,5,6]
>>> if None not in (a,b):
...     print(a,b)
...
[4, 5, 6] [1, 2, 3]
>>> if not None in (a,b):
...     print(a,b)
...
[4, 5, 6] [1, 2, 3]
>>>


>>> a = None
>>> if not None in (a,b):
...     print(a,b)
...
>>> if None not in (a,b):
...     print(a,b)
...
>>>
`

Reply:
True, they're equivalent. I usually write `x not in y` because it's more clear. If you write not y in x it might be unclear whether you meant not (y in x) or (not y) in x. No need to change though. Just a stylistic nit.